### PR TITLE
DAOS-623 cq: Move shellcheck tests to GitHub actions.

### PR DESCRIPTION
**Pull request by ashleypittman**
Add a new linter check for shellcheck.
Silence many warnings, making use of new -P option.
Remove check from Jenkins.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: .github/workflows/linting.yml [hunk 1]
```diff
@@ -30,6 +30,24 @@ jobs:
...
+          ref: ${{ github.event.pull_request.head.sha }}
+      - name: Run
+        run: sudo apt-get update && sudo apt-get install shellcheck
+      - name: Add error parser
+        run: echo -n "::add-matcher::ci/shellcheck-matcher.json"
+      - name: Run Shellcheck
+        # The check will run with this file from the target branch but the code from the PR so
+        # test for this file before calling it to prevent failures on PRs where this check is
+        # in the target branch but the PR is not updated to include it.
+        run: \[ ! -x ci/run_shellcheck.sh \] || ./ci/run_shellcheck.sh
```
Comment (brianjmurrell) [inline .github/workflows/linting.yml:49]: Surely there is an action available already to do all of this, no?

File: src/tests/ftest/scripts/main.sh [hunk 2]
```diff
@@ -186,7 +186,7 @@ if [ "${STAGE_NAME}" == "Functional Hardware 24" ]; then
 fi
 
 # shellcheck disable=SC2086,SC2090
-if ! ./launch.py --mode ci ${launch_node_args} ${LAUNCH_OPT_ARGS} ${TEST_TAG_ARR[*]}; then
+if ! ./launch.py --mode ci ${launch_node_args} ${LAUNCH_OPT_ARGS} "${TEST_TAG_ARR[*]}"; then
```
Comment (brianjmurrell) [inline src/tests/ftest/scripts/main.sh:190]: You can't quote that.  I have fallen into that trap several times.

```suggestion
if ! ./launch.py --mode ci ${launch_node_args} ${LAUNCH_OPT_ARGS} ${TEST_TAG_ARR[*]}; then
```

File: utils/docker/vcluster/daos-cm.sh [hunk 3]
```diff
@@ -231,7 +234,7 @@ done
 CMD="$1"
 DAOS_IFACE_IP="$2"
 
-cd "$CWD"
+cd "$CWD" || exit 1
```
Comment (brianjmurrell) [inline utils/docker/vcluster/daos-cm.sh:237]: `set -e` would be much more encompassing and thorough.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: src/tests/ftest/scripts/main.sh (hunk 2 above)
Comment (ashleypittman) [reply src/tests/ftest/scripts/main.sh:190]: I'll just suppress it then as it's clearly working for now.

File: utils/docker/vcluster/daos-cm.sh (hunk 3 above)
Comment (ashleypittman) [reply utils/docker/vcluster/daos-cm.sh:237]: I agree but without being able to run/test this code I didn't want to do that.  @knard-intel any thoughts here?

Comment (knard-intel) [reply utils/docker/vcluster/daos-cm.sh:237]: After reviewing the code, I do not remember why I do not have put `-e` with the `-o pipefail`.
Seems OK for me to put the `-e` for me.
I am working on a PR where I will manually test this script.
In conclusion, using the `-e` seems OK for me.

File: src/tests/ftest/scripts/main.sh (hunk 2 above)
Comment (knard-intel) [reply src/tests/ftest/scripts/main.sh]: I think there is a issue on the name of the variable:
s/TEST_TAG_ARR/TEST_TAG_ARG/

Comment (ashleypittman) [reply src/tests/ftest/scripts/main.sh]: I think TEST_TAG_ARG is a string of strings, TEST_TAG_ARR is an array of strings from TEST_TAG_ARG.

File: utils/docker/vcluster/daos-cm.sh (hunk 3 above)
Comment (ashleypittman) [reply utils/docker/vcluster/daos-cm.sh]: I had to suppress a couple of things in this directory I'd prefer not to have done, if you are able to take a closer look when doing future work in this area that would be great.

File: utils/sl/setup_local.sh [hunk 4]
```diff
@@ -105,6 +101,7 @@ for item in $list; do
     fi
 done
 
+# shellcheck disable=SC2181
 in_list "${SL_PREFIX}/bin" "${old_path}"
 if [ $? -eq 0 ]; then
```
Comment (brianjmurrell) [inline utils/sl/setup_local.sh:106]: What was wrong with your previously proposed
```suggestion
if ! in_list "${SL_PREFIX}/bin" "${old_path}"
```

This is the more idiomatic syntax.

Comment (ashleypittman) [reply utils/sl/setup_local.sh:106]: It wasn't working, there was a unit test failure as a result of this change.

File: Jenkinsfile [hunk 5]
```diff
@@ -401,66 +401,6 @@ pipeline {
...
-                    agent {
-                        dockerfile {
-                            filename 'Dockerfile.checkpatch'
-                            dir 'utils/docker'
-                            label 'docker_runner'
-                            additionalBuildArgs dockerBuildArgs(add_repos: false)
-                        }
-                    }
-                    steps {
-                        checkPatch user: GITHUB_USER_USR,
```
Comment (brianjmurrell) [inline Jenkinsfile:418]: Are we really ready to remove checkpatch entirely just be putting shellcheck into GHA?

File: ci/run_shellcheck.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+set -e
+
+# Find and run shellcheck on all shell scripts.  Previously the `file` command was used
+# to identify shell commands but that checks for a #! which some of our shell code does not
+# use so go purely on filename.
+# This depends on shellcheck 0.7.2 or above so works in GitHub actions but not on el8.8
+find . \( -path ./.git -o -path ./venv -o -path ./build -o -path ./src/control/vendor \
+ -o -path ./install -o -path ./src/rdb/raft -type d  \) -prune -o -name "*.sh" -exec \
+ shellcheck -P ci --external -f gcc \{\} \+
```
Comment (brianjmurrell) [inline ci/run_shellcheck.sh:11]: Is the argument parser doing some kind of closest-match or prefix matching or something?  https://github.com/koalaman/shellcheck/blob/master/shellcheck.1.md documents the argument as `-x, --external-sources`

File: src/client/java/daos-java/find_java_home.sh [hunk 7]
```diff
@@ -5,6 +5,7 @@ if [ -n  "$java_home_path" ]
 then
 	echo "$java_home_path"
 else
+	# shellcheck disable=SC2046
 	java_path="$(dirname $(readlink -f $(which java)))"
```
Comment (brianjmurrell) [inline src/client/java/daos-java/find_java_home.sh:9]: I think we just want to properly quote here:

```suggestion
	# shellcheck disable=SC2046
	java_path="$(dirname "$(readlink -f "$(command -v java)")")"
```

(`command -v` to avoid [SC2230](https://www.shellcheck.net/wiki/SC2230) which I'm surprised we are not getting an error for).

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 5 above)
Comment (ashleypittman) [reply Jenkinsfile:418]: yes!

File: src/cart/multi-node-test.sh [hunk 8]
```diff
@@ -1,196 +0,0 @@
-#!/bin/bash
```
Comment (daltonbohning) [inline src/cart/multi-node-test.sh:1]: Why was this file removed?

File: Jenkinsfile (hunk 5 above)
Comment (daltonbohning) [reply Jenkinsfile:418]: I think it's great to do this, but will it introduce more disparity between the master and 2.4 checks if not backported?

File: src/cart/multi-node-test.sh (hunk 8 above)
Comment (ashleypittman) [reply src/cart/multi-node-test.sh:1]: I think it's stale and not doing anything.

File: Jenkinsfile (hunk 5 above)
Comment (ashleypittman) [reply Jenkinsfile:418]: That's inevitable.

We will have a master that is shellcheck-clean but release branches that still has issues.

File: ci/run_shellcheck.sh (hunk 6 above)
Comment (ashleypittman) [reply ci/run_shellcheck.sh:11]: Ah, probably that.  I went through a few iterations here but this came from https://github.com/daos-stack/code_review/blob/master/shellcheck_scripts.sh#L35 which also uses `--external`.  Will expand all these options if I need to re-push.

File: src/client/java/daos-java/find_java_home.sh (hunk 7 above)
Comment (ashleypittman) [reply src/client/java/daos-java/find_java_home.sh:9]: You're probably right but again, I'm not familiar with this code or how much testing it gets so I'd like to make as few changes as possible.

File: src/client/java/hadoop-daos/src/test/resources/java-test.sh [hunk 9]
```diff
@@ -1,10 +1,12 @@
 #!/bin/sh
 
+set -e
 cd /root/daos
 first=$1
 shift
 caller=$1
 
+# shellcheck disable=SC3010
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/java/hadoop-daos/src/test/resources/java-test.sh:9]: Probably.  I'll raise this next time someone is working on the java code.

File: Jenkinsfile (hunk 5 above)
Comment (ashleypittman) [reply Jenkinsfile:418]: There are two models for dealing with warnings, either we accept we have them and warn but do not fail PRs where files with warnings are modified.  That's what we used to do with pylint and we do today with shellcheck on both master and the release branch.  The other model is the tree is clean, we check the whole tree on every PR and fail the build on warrnings.

At this point there is always going to be a release build, to not have disparity between branches then we'd have to fix all the issues on master, fix all the issue on the release branch, then upgrade the test on master and then release in the mean time not landing any regressions to master.

Clearly possible but a lot more work and for the small number of files affected is it worth it?

Shellcheck is also the last check (that I'm aware of) where we only check the diff, other than the clang-format change which is a lot newer and and differs there because most code isn't compliant.

File: src/client/java/daos-java/find_java_home.sh (hunk 7 above)
Comment (knard-intel) [reply src/client/java/daos-java/find_java_home.sh:9]: I agree with @brianjmurrell fix proposal as I have experimented errors with missing quoting and the fix seems pretty safe.
I really think that it is an error or forget from the author.

File: src/tests/ftest/scripts/main.sh [hunk 10]
```diff
@@ -1,4 +1,5 @@
 # shellcheck disable=SC1113
+# shellcheck shell=bash
```
Comment (knard-intel) [inline src/tests/ftest/scripts/main.sh:2]: Maybe we could also add the #!/bin/bash header ?

File: src/tests/ftest/scripts/setup_nodes.sh [hunk 11]
```diff
@@ -1,4 +1,5 @@
 # shellcheck disable=SC1113
+# shellcheck shell=bash
```
Comment (knard-intel) [inline src/tests/ftest/scripts/setup_nodes.sh:2]: Maybe we could also add the #!/bin/bash header ?

File: utils/docker/examples/client/deploy-docker_stack.sh [hunk 12]
```diff
@@ -3,9 +3,11 @@
 # set -x
 set -e -o pipefail
 
+# shellcheck disable=SC2086
 CWD="$(realpath "$(dirname $0)")"
```
[the remark this answers is not in the document]
Comment (knard-intel) [reply utils/docker/examples/client/deploy-docker_stack.sh:7]: Agree on this patch proposal

File: utils/docker/vcluster/daos-cm.sh [hunk 13]
```diff
@@ -1,8 +1,11 @@
 #!/bin/bash
 
-set -o pipefail
+# shellcheck disable=SC2034,SC2145,SC2086,SC2068
+
+set -e -o pipefail
 
 VERSION=0.2
+# shellcheck disable=SC2046
 CWD="$(realpath $(dirname $0))"
```
[the remark this answers is not in the document]
Comment (knard-intel) [reply utils/docker/vcluster/daos-cm.sh:9]: Agree on this proposal

File: utils/docker/vcluster/daos-cm.sh [hunk 14]
```diff
@@ -1,8 +1,11 @@
 #!/bin/bash
 
-set -o pipefail
+# shellcheck disable=SC2034,SC2145,SC2086,SC2068
```
[the remark this answers is not in the document]
Comment (knard-intel) [reply utils/docker/vcluster/daos-cm.sh:3]: I have a draft PR https://github.com/daos-stack/daos/pull/13087 working on this file.
I could add the fixes of the shellcheck in it.

**Review by knard-intel — APPROVED** [review]
LGTM

File: utils/sl/setup_local.sh (hunk 4 above)
Comment (brianjmurrell) [reply utils/sl/setup_local.sh:106]: It's odd that it wasn't working.  They are functionally the equivalent when `set -e` isn't being used (which is a bad thing in it's own right) and the existing syntax won't work if `set -e` is being used.

So `set -e` not being in use here is another issue.

Comment (ashleypittman) [reply utils/sl/setup_local.sh:106]: This script is intended to be sourced by the user to setup their env.  If you use set -e or exit then it logs the user out which is not the desired behaviour.

File: utils/docker/examples/client/deploy-docker_stack.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply utils/docker/examples/client/deploy-docker_stack.sh]: Although, to be thorough, the elimination of `dirname` can be done here also:

```suggestion
CWD="$(realpath "${0%/*}")"
```

**Review by jolivier23 — APPROVED** [review]

File: utils/docker/vcluster/daos-cm.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (ashleypittman) [reply utils/docker/vcluster/daos-cm.sh:9]: It looks like this is the only remaining requested change, if that's the case can we move forward with this PR and have @knard-intel fix this file when he's working on it.

[the remark this answers is not in the document]
Comment (knard-intel) [reply utils/docker/vcluster/daos-cm.sh:9]: As indicated by @ashleypittman , it will be fixed with the PR https://github.com/daos-stack/daos/pull/13087/files#diff-7d867865cdddd1f9e49de143c074014d98a25c41456c373544d3857493be2a29

**Comment by ashleypittman** [comment]
@daos-stack/daos-gatekeeper this has previously passed all relevant tests, I merged with master today and the code that it modifies and the test that it adds all pass.

**Comment by phender** [comment]
The `daos_test/rebuild.py` failures in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13360/21/ are due to the tests being unable to start the servers because of a missing certificate.  The certificate was missing because of a failed copy:
```
2023-12-10 19:14:50,422 command_utils    L1005 DEBUG| Copying certificates for dmg:
2023-12-10 19:14:50,793 command_utils    L1014 DEBUG|   /var/tmp/daos_testing/daosCA/certs/daosCA.crt -> /etc/daos/certs/daosCA.crt
2023-12-10 19:14:51,550 command_utils    L1014 DEBUG|   /var/tmp/daos_testing/daosCA/certs/admin.crt -> /etc/daos/certs/admin.crt
2023-12-10 19:14:51,928 command_utils    L1022 INFO |     WARNING: /etc/daos/certs/admin.crt copy failed on wolf-253:
command: 'clush -S -v -w wolf-253 sudo -n cp /var/tmp/daos_testing/daosCA/certs/admin.crt /etc/daos/certs/admin.crt'
exit_status: 1
duration: 0.37566280364990234
interrupted: False
pid: 44794
encoding: 'UTF-8'
stdout: b''
stderr: b"Warning: fd_max set to 8192 but max open files hard limit is 4096\nwolf-253: cp: cannot stat '/var/tmp/daos_testing/daosCA/certs/admin.crt': No such file or directory\nclush: wolf-253: exited with exit code 1\n"
```
The reason the wolf-253:/var/tmp/daos_testing/daosCA/certs directory does exist is because the `gen_certificates.sh ` command failed:
```
2023/12/10 07:14:24 DEBUG                      run_local: Running on wolf-253: /usr/lib64/daos/certgen/gen_certificates.sh /var/tmp/daos_testing
2023/12/10 07:14:25 DEBUG                      run_local:   wolf-253 (rc=0):
2023/12/10 07:14:25 DEBUG                      run_local:     Generating Private CA Root Certificate
2023/12/10 07:14:25 DEBUG                      run_local:     Generating RSA private key, 3072 bit long modulus (2 primes)
2023/12/10 07:14:25 DEBUG                      run_local:     ................................................++++
2023/12/10 07:14:25 DEBUG                      run_local:     .....................................++++
2023/12/10 07:14:25 DEBUG                      run_local:     e is 65537 (0x010001)
2023/12/10 07:14:25 DEBUG                      run_local:     Private CA Root Certificate created in /var/tmp/daos_testing/daosCA
2023/12/10 07:14:25 DEBUG                      run_local:     Generating Server Certificate
2023/12/10 07:14:25 DEBUG                      run_local:     Generating RSA private key, 3072 bit long modulus (2 primes)
2023/12/10 07:14:25 DEBUG                      run_local:     ............................................................................++++
2023/12/10 07:14:25 DEBUG                      run_local:     ...............................++++
2023/12/10 07:14:25 DEBUG                      run_local:     e is 65537 (0x010001)
2023/12/10 07:14:25 DEBUG                      run_local:     Can't open /usr/lib64/daos/server.cnf for reading, No such file or directory
2023/12/10 07:14:25 DEBUG                      run_local:     140565353428800:error:02001002:system library:fopen:No such file or directory:crypto/bio/bss_file.c:69:fopen('/usr/lib64/daos/server.cnf','r')
2023/12/10 07:14:25 DEBUG                      run_local:     140565353428800:error:2006D080:BIO routines:BIO_new_file:no such file:crypto/bio/bss_file.c:76:
2023/12/10 07:14:25 DEBUG                      run_local:     Using configuration from /var/tmp/daos_testing/daosCA/ca.cnf
2023/12/10 07:14:25 DEBUG                      run_local:     140580365395776:error:02001002:system library:fopen:No such file or directory:crypto/bio/bss_file.c:69:fopen('/var/tmp/daos_testing/daosCA/server.csr','r')
2023/12/10 07:14:25 DEBUG                      run_local:     140580365395776:error:2006D080:BIO routines:BIO_new_file:no such file:crypto/bio/bss_file.c:76:
2023/12/10 07:14:25 DEBUG                      run_local:     chmod: cannot access '/var/tmp/daos_testing/daosCA/certs/server.crt': No such file or directory
2023/12/10 07:14:25 DEBUG                      run_local:     Required Server Certificate Files:
2023/12/10 07:14:25 DEBUG                      run_local:         /var/tmp/daos_testing/daosCA/certs/daosCA.crt
2023/12/10 07:14:25 DEBUG                      run_local:         /var/tmp/daos_testing/daosCA/certs/server.key
2023/12/10 07:14:25 DEBUG                      run_local:         /var/tmp/daos_testing/daosCA/certs/server.crt
2023/12/10 07:14:25 DEBUG                      run_local:     Generating Agent Certificate
2023/12/10 07:14:25 DEBUG                      run_local:     Generating RSA private key, 3072 bit long modulus (2 primes)
```

File: utils/certs/gen_certificates.sh [hunk 15]
```diff
@@ -22,7 +22,7 @@ CA_HOME="${1:-.}/daosCA"
 PRIVATE="${CA_HOME}/private"
 CERTS="${CA_HOME}/certs"
 CLIENTS="${CERTS}/clients"
-CONFIGS="$(dirname "${BASH_SOURCE[0]%/*}")"
+CONFIGS="$(dirname "${BASH_SOURCE}")"
```
Comment (brianjmurrell) [inline utils/certs/gen_certificates.sh:26]: This only needed reverting because my [original suggestion](https://github.com/daos-stack/daos/pull/13360/files/88eb485b65e6a41053c0e94bfe18d7396d89d7b4#r1399504327) was incorrect but I [did correct it subsequently](https://github.com/daos-stack/daos/pull/13360/files/88eb485b65e6a41053c0e94bfe18d7396d89d7b4#r1406477480) but that suggestion was never implemented.

File: utils/sl/setup_local.sh (hunk 4 above)
Comment (ashleypittman) [reply utils/sl/setup_local.sh:106]: Neither do I, and yet it was the only change in that build.  Like I say my goal is not to fix or work on the existing warnings but improve our processes so that we don't end up with more in the codebase.  It would be well worth a ticket to resolve all currently suppressed shellcheck warnings.

File: utils/docker/examples/client/deploy-docker_stack.sh [hunk 16]
```diff
@@ -3,8 +3,7 @@
 # set -x
 set -e -o pipefail
 
-# shellcheck disable=SC2086
-CWD="$(realpath "$(dirname $0)")"
+CWD="$(realpath "$(dirname "$0")")"
```
[the remark this answers is not in the document]
Comment (knard-intel) [reply utils/docker/examples/client/deploy-docker_stack.sh:6]: fixed with commit [7a87891](https://github.com/daos-stack/daos/pull/13076/commits/7a878910e3aeb241265270d8c66ae5c231c9de00)

File: utils/docker/vcluster/daos-cm.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (knard-intel) [reply utils/docker/vcluster/daos-cm.sh:9]: New version of the CWD fix could be find at https://github.com/daos-stack/daos/pull/13087/files#diff-7d867865cdddd1f9e49de143c074014d98a25c41456c373544d3857493be2a29

**Review by knard-intel — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]
